gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java

```diff
@@ -0,0 +1,273 @@
+    this.jobResourceName = TaskUtil.getNamespacedJobName(this.helixQueueName, this.jobContext.getJobId());
+
+    this.stateSerDeRunnerThreads = Integer.parseInt(jobProps.getProperty(ParallelRunner.PARALLEL_RUNNER_THREADS_KEY,
+        Integer.toString(ParallelRunner.DEFAULT_PARALLEL_RUNNER_THREADS)));
+  }
+
+  @Override
+  protected void runWorkUnits(List<WorkUnit> workUnits) throws Exception {
+    try {
+      submitJobToHelix(this.jobContext.getJobName(), this.jobContext.getJobId(),
+          createJob(this.jobContext.getJobName(), this.jobContext.getJobId(), workUnits));
+      waitForJobCompletion();
+      this.jobContext.getJobState().setState(JobState.RunningState.SUCCESSFUL);
+      this.jobContext.getJobState().addTaskStates(collectOutputTaskStates());
+    } finally {
+      deletePersistedWorkUnitsForJob(this.jobContext.getJobId());
+    }
+  }
+
+  @Override
+  protected JobLock getJobLock() throws IOException {
+    return new FileBasedJobLock(this.fs, this.jobProps.getProperty(ConfigurationKeys.JOB_LOCK_DIR_KEY),
+        this.jobContext.getJobName());
+  }
+
+  @Override
+  protected void executeCancellation() {
+    // Currently not supported yet
+  }
+
+  /**
+   * Create a job from a given batch of {@link WorkUnit}s.
+   */
+  private JobConfig.Builder createJob(String jobName, String jobId, List<WorkUnit> workUnits)
+      throws IOException {
+    Map<String, TaskConfig> taskConfigMap = Maps.newHashMap();
+
+    Closer closer = Closer.create();
+    try {
+      ParallelRunner stateSerDeRunner = closer.register(new ParallelRunner(this.stateSerDeRunnerThreads, this.fs));
+
+      for (WorkUnit workUnit : workUnits) {
+        if (workUnit instanceof MultiWorkUnit) {
+          // Flatten the MultiWorkUnit and add each individual WorkUnit
+          for (WorkUnit innerWorkUnit : JobLauncherUtils.flattenWorkUnits(((MultiWorkUnit) workUnit).getWorkUnits())) {
+            addWorkUnit(innerWorkUnit, stateSerDeRunner, taskConfigMap, jobName, jobId);
+          }
+        } else {
+          addWorkUnit(workUnit, stateSerDeRunner, taskConfigMap, jobName, jobId);
+        }
+      }
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+
+    JobConfig.Builder jobConfigBuilder = new JobConfig.Builder();
+    jobConfigBuilder.addTaskConfigMap(taskConfigMap).setCommand(GobblinWorkUnitRunner.GOBBLIN_TASK_FACTORY_NAME);
+
+    return jobConfigBuilder;
+  }
+
+  /**
+   * Submit a job to run.
+   */
+  private void submitJobToHelix(String jobName, String jobId, JobConfig.Builder jobConfigBuilder) throws Exception {
+    // Create one queue for each job with the job name being the queue name
+    JobQueue jobQueue = new JobQueue.Builder(jobName).build();
+    try {
+      this.helixTaskDriver.createQueue(jobQueue);
+    } catch (IllegalArgumentException iae) {
+      LOGGER.warn(String.format("Job queue %s already exists", jobQueue.getName()));
```

Comment:
This could be an `info` rather than `warn` since if a job runs multiple times then it's normal that from the 2nd time the jobQueue will already exist.